appendExpr(): Improve handling of operators

cassella opened the pull request:
Handle prefix-~ and postfix-!

If the two-actual expr's name starts with a letter (or _), don't print
it as an operator.  Treat it the same as an expr with other than two
actuals.

Also tidy the logic so that two-arg non-operators use the same else
block as not-two-arg cases.

Add three tests locking in the changed behavior.

Add futures for #16939, #16940, and #16941, done in this PR since they depend on files added for one of the above tests.

Resolves #13626

cassella commented:
I'm not super confident in this approach, but `tests/chpldoc` all pass, and the resulting changes in `make docs` output all look like improvements.

Suggestions for more confident-sounding comments would be appreciated.

`diff -u doc-master doc-pr` (excluding searchindex.js)
<details>
<pre>
diff -r doc-master/html/modules/packages/HDF5.html doc-pr/html/modules/packages/HDF5.html
329c329
< <em class="property">proc </em><code class="sig-name descname">writeArraysToHDF5Files</code><span class="sig-paren">(</span><em class="sig-param">dirName: string</em>, <em class="sig-param">dsetNames: [] string</em>, <em class="sig-param">filenames: [] string</em>, <em class="sig-param">type eltType</em>, <em class="sig-param">param rank: int</em>, <em class="sig-param">data: [] eltTypeArrayWrapperrank</em>, <em class="sig-param">mode: Hdf5OpenMode</em><span class="sig-paren">)</span> throws<a class="headerlink" href="#HDF5.writeArraysToHDF5Files" title="Permalink to this definition">¶</a></dt>
---
> <em class="property">proc </em><code class="sig-name descname">writeArraysToHDF5Files</code><span class="sig-paren">(</span><em class="sig-param">dirName: string</em>, <em class="sig-param">dsetNames: [] string</em>, <em class="sig-param">filenames: [] string</em>, <em class="sig-param">type eltType</em>, <em class="sig-param">param rank: int</em>, <em class="sig-param">data: [] ArrayWrapper(eltType</em>, <em class="sig-param">rank)</em>, <em class="sig-param">mode: Hdf5OpenMode</em><span class="sig-paren">)</span> throws<a class="headerlink" href="#HDF5.writeArraysToHDF5Files" title="Permalink to this definition">¶</a></dt>
diff -r doc-master/html/modules/packages/OrderedSet.html doc-pr/html/modules/packages/OrderedSet.html
502c502
< <em class="property">proc </em><code class="sig-name descname">isDisjoint</code><span class="sig-paren">(</span><em class="sig-param">const ref other: eltTypeorderedSet?</em><span class="sig-paren">)</span>: bool<a class="headerlink" href="#OrderedSet.orderedSet.isDisjoint" title="Permalink to this definition">¶</a></dt>
---
> <em class="property">proc </em><code class="sig-name descname">isDisjoint</code><span class="sig-paren">(</span><em class="sig-param">const ref other: orderedSet(eltType</em>, <em class="sig-param">?)</em><span class="sig-paren">)</span>: bool<a class="headerlink" href="#OrderedSet.orderedSet.isDisjoint" title="Permalink to this definition">¶</a></dt>
520c520
< <em class="property">proc </em><code class="sig-name descname">isIntersecting</code><span class="sig-paren">(</span><em class="sig-param">const ref other: eltTypeorderedSet?</em><span class="sig-paren">)</span>: bool<a class="headerlink" href="#OrderedSet.orderedSet.isIntersecting" title="Permalink to this definition">¶</a></dt>
---
> <em class="property">proc </em><code class="sig-name descname">isIntersecting</code><span class="sig-paren">(</span><em class="sig-param">const ref other: orderedSet(eltType</em>, <em class="sig-param">?)</em><span class="sig-paren">)</span>: bool<a class="headerlink" href="#OrderedSet.orderedSet.isIntersecting" title="Permalink to this definition">¶</a></dt>
714c714
< <code class="sig-name descname">proc &amp;=(ref lhs: ?torderedSet?, const ref rhs: torderedSet?)</code></dt>
---
> <code class="sig-name descname">proc &amp;=(ref lhs: orderedSet(?t, ?), const ref rhs: orderedSet(t, ?))</code></dt>
diff -r doc-master/html/modules/standard/Heap.html doc-pr/html/modules/standard/Heap.html
406c406
< <em class="property">proc </em><code class="sig-name descname">=</code><span class="sig-paren">(</span><em class="sig-param">ref lhs: ?theap?</em>, <em class="sig-param">ref rhs: theap?</em><span class="sig-paren">)</span><a class="headerlink" href="#Heap.=" title="Permalink to this definition">¶</a></dt>
---
> <em class="property">proc </em><code class="sig-name descname">=</code><span class="sig-paren">(</span><em class="sig-param">ref lhs: heap(?t</em>, <em class="sig-param">?)</em>, <em class="sig-param">ref rhs: heap(t</em>, <em class="sig-param">?)</em><span class="sig-paren">)</span><a class="headerlink" href="#Heap.=" title="Permalink to this definition">¶</a></dt>
diff -r doc-master/html/modules/standard/List.html doc-pr/html/modules/standard/List.html
477c477
< <code class="sig-name descname">proc ref extend(other: eltTypelist?p)</code></dt>
---
> <code class="sig-name descname">proc ref extend(other: list(eltType, ?p))</code></dt>
951c951
< <em class="property">proc </em><code class="sig-name descname">=</code><span class="sig-paren">(</span><em class="sig-param">ref lhs: ?tlist?</em>, <em class="sig-param">rhs: tlist?</em><span class="sig-paren">)</span><a class="headerlink" href="#List.=" title="Permalink to this definition">¶</a></dt>
---
> <em class="property">proc </em><code class="sig-name descname">=</code><span class="sig-paren">(</span><em class="sig-param">ref lhs: list(?t</em>, <em class="sig-param">?)</em>, <em class="sig-param">rhs: list(t</em>, <em class="sig-param">?)</em><span class="sig-paren">)</span><a class="headerlink" href="#List.=" title="Permalink to this definition">¶</a></dt>
971c971
< <em class="property">proc </em><code class="sig-name descname">==</code><span class="sig-paren">(</span><em class="sig-param">a: ?tlist?</em>, <em class="sig-param">b: tlist?</em><span class="sig-paren">)</span>: bool<a class="headerlink" href="#List.==" title="Permalink to this definition">¶</a></dt>
---
> <em class="property">proc </em><code class="sig-name descname">==</code><span class="sig-paren">(</span><em class="sig-param">a: list(?t</em>, <em class="sig-param">?)</em>, <em class="sig-param">b: list(t</em>, <em class="sig-param">?)</em><span class="sig-paren">)</span>: bool<a class="headerlink" href="#List.==" title="Permalink to this definition">¶</a></dt>
991c991
< <em class="property">proc </em><code class="sig-name descname">!=</code><span class="sig-paren">(</span><em class="sig-param">a: ?tlist?</em>, <em class="sig-param">b: tlist?</em><span class="sig-paren">)</span>: bool<a class="headerlink" href="#List.!=" title="Permalink to this definition">¶</a></dt>
---
> <em class="property">proc </em><code class="sig-name descname">!=</code><span class="sig-paren">(</span><em class="sig-param">a: list(?t</em>, <em class="sig-param">?)</em>, <em class="sig-param">b: list(t</em>, <em class="sig-param">?)</em><span class="sig-paren">)</span>: bool<a class="headerlink" href="#List.!=" title="Permalink to this definition">¶</a></dt>
diff -r doc-master/html/modules/standard/Set.html doc-pr/html/modules/standard/Set.html
289c289
< <em class="property">proc </em><code class="sig-name descname">init=</code><span class="sig-paren">(</span><em class="sig-param">const ref other: ?tset?</em><span class="sig-paren">)</span><a class="headerlink" href="#Set.set.init=" title="Permalink to this definition">¶</a></dt>
---
> <em class="property">proc </em><code class="sig-name descname">init=</code><span class="sig-paren">(</span><em class="sig-param">const ref other: set(?t</em>, <em class="sig-param">?)</em><span class="sig-paren">)</span><a class="headerlink" href="#Set.set.init=" title="Permalink to this definition">¶</a></dt>
332c332
< <em class="property">proc </em><code class="sig-name descname">isDisjoint</code><span class="sig-paren">(</span><em class="sig-param">const ref other: eltTypeset?</em><span class="sig-paren">)</span>: bool<a class="headerlink" href="#Set.set.isDisjoint" title="Permalink to this definition">¶</a></dt>
---
> <em class="property">proc </em><code class="sig-name descname">isDisjoint</code><span class="sig-paren">(</span><em class="sig-param">const ref other: set(eltType</em>, <em class="sig-param">?)</em><span class="sig-paren">)</span>: bool<a class="headerlink" href="#Set.set.isDisjoint" title="Permalink to this definition">¶</a></dt>
354c354
< <em class="property">proc </em><code class="sig-name descname">isIntersecting</code><span class="sig-paren">(</span><em class="sig-param">const ref other: eltTypeset?</em><span class="sig-paren">)</span>: bool<a class="headerlink" href="#Set.set.isIntersecting" title="Permalink to this definition">¶</a></dt>
---
> <em class="property">proc </em><code class="sig-name descname">isIntersecting</code><span class="sig-paren">(</span><em class="sig-param">const ref other: set(eltType</em>, <em class="sig-param">?)</em><span class="sig-paren">)</span>: bool<a class="headerlink" href="#Set.set.isIntersecting" title="Permalink to this definition">¶</a></dt>
474c474
< <em class="property">proc </em><code class="sig-name descname">=</code><span class="sig-paren">(</span><em class="sig-param">ref lhs: ?tset?</em>, <em class="sig-param">const ref rhs: tset?</em><span class="sig-paren">)</span><a class="headerlink" href="#Set.=" title="Permalink to this definition">¶</a></dt>
---
> <em class="property">proc </em><code class="sig-name descname">=</code><span class="sig-paren">(</span><em class="sig-param">ref lhs: set(?t</em>, <em class="sig-param">?)</em>, <em class="sig-param">const ref rhs: set(t</em>, <em class="sig-param">?)</em><span class="sig-paren">)</span><a class="headerlink" href="#Set.=" title="Permalink to this definition">¶</a></dt>
494c494
< <code class="sig-name descname">proc |(const ref a: ?tset?, const ref b: tset?): set(t)</code></dt>
---
> <code class="sig-name descname">proc |(const ref a: set(?t, ?), const ref b: set(t, ?)): set(t)</code></dt>
514c514
< <code class="sig-name descname">proc |=(ref lhs: ?tset?, const ref rhs: tset?)</code></dt>
---
> <code class="sig-name descname">proc |=(ref lhs: set(?t, ?), const ref rhs: set(t, ?))</code></dt>
528c528
< <em class="property">proc </em><code class="sig-name descname">+</code><span class="sig-paren">(</span><em class="sig-param">const ref a: ?tset?</em>, <em class="sig-param">const ref b: tset?</em><span class="sig-paren">)</span>: tset?<a class="headerlink" href="#Set.+" title="Permalink to this definition">¶</a></dt>
---
> <em class="property">proc </em><code class="sig-name descname">+</code><span class="sig-paren">(</span><em class="sig-param">const ref a: set(?t</em>, <em class="sig-param">?)</em>, <em class="sig-param">const ref b: set(t</em>, <em class="sig-param">?)</em><span class="sig-paren">)</span>: set(t, ?)<a class="headerlink" href="#Set.+" title="Permalink to this definition">¶</a></dt>
549c549
< <em class="property">proc </em><code class="sig-name descname">+=</code><span class="sig-paren">(</span><em class="sig-param">ref lhs: ?tset?</em>, <em class="sig-param">const ref rhs: tset?</em><span class="sig-paren">)</span><a class="headerlink" href="#Set.+=" title="Permalink to this definition">¶</a></dt>
---
> <em class="property">proc </em><code class="sig-name descname">+=</code><span class="sig-paren">(</span><em class="sig-param">ref lhs: set(?t</em>, <em class="sig-param">?)</em>, <em class="sig-param">const ref rhs: set(t</em>, <em class="sig-param">?)</em><span class="sig-paren">)</span><a class="headerlink" href="#Set.+=" title="Permalink to this definition">¶</a></dt>
563c563
< <em class="property">proc </em><code class="sig-name descname">-</code><span class="sig-paren">(</span><em class="sig-param">const ref a: ?tset?</em>, <em class="sig-param">const ref b: tset?</em><span class="sig-paren">)</span>: set(t)<a class="headerlink" href="#Set.-" title="Permalink to this definition">¶</a></dt>
---
> <em class="property">proc </em><code class="sig-name descname">-</code><span class="sig-paren">(</span><em class="sig-param">const ref a: set(?t</em>, <em class="sig-param">?)</em>, <em class="sig-param">const ref b: set(t</em>, <em class="sig-param">?)</em><span class="sig-paren">)</span>: set(t)<a class="headerlink" href="#Set.-" title="Permalink to this definition">¶</a></dt>
583c583
< <em class="property">proc </em><code class="sig-name descname">-=</code><span class="sig-paren">(</span><em class="sig-param">ref lhs: ?tset?</em>, <em class="sig-param">const ref rhs: tset?</em><span class="sig-paren">)</span><a class="headerlink" href="#Set.-=" title="Permalink to this definition">¶</a></dt>
---
> <em class="property">proc </em><code class="sig-name descname">-=</code><span class="sig-paren">(</span><em class="sig-param">ref lhs: set(?t</em>, <em class="sig-param">?)</em>, <em class="sig-param">const ref rhs: set(t</em>, <em class="sig-param">?)</em><span class="sig-paren">)</span><a class="headerlink" href="#Set.-=" title="Permalink to this definition">¶</a></dt>
602c602
< <code class="sig-name descname">proc &amp;(const ref a: ?tset?, const ref b: tset?): set(t)</code></dt>
---
> <code class="sig-name descname">proc &amp;(const ref a: set(?t, ?), const ref b: set(t, ?)): set(t)</code></dt>
622c622
< <code class="sig-name descname">proc &amp;=(ref lhs: ?tset?, const ref rhs: tset?)</code></dt>
---
> <code class="sig-name descname">proc &amp;=(ref lhs: set(?t, ?), const ref rhs: set(t, ?))</code></dt>
642c642
< <code class="sig-name descname">proc ^(const ref a: ?tset?, const ref b: tset?): set(t)</code></dt>
---
> <code class="sig-name descname">proc ^(const ref a: set(?t, ?), const ref b: set(t, ?)): set(t)</code></dt>
662c662
< <code class="sig-name descname">proc ^=(ref lhs: ?tset?, const ref rhs: tset?)</code></dt>
---
> <code class="sig-name descname">proc ^=(ref lhs: set(?t, ?), const ref rhs: set(t, ?))</code></dt>
682c682
< <em class="property">proc </em><code class="sig-name descname">==</code><span class="sig-paren">(</span><em class="sig-param">const ref a: ?tset?</em>, <em class="sig-param">const ref b: tset?</em><span class="sig-paren">)</span>: bool<a class="headerlink" href="#Set.==" title="Permalink to this definition">¶</a></dt>
---
> <em class="property">proc </em><code class="sig-name descname">==</code><span class="sig-paren">(</span><em class="sig-param">const ref a: set(?t</em>, <em class="sig-param">?)</em>, <em class="sig-param">const ref b: set(t</em>, <em class="sig-param">?)</em><span class="sig-paren">)</span>: bool<a class="headerlink" href="#Set.==" title="Permalink to this definition">¶</a></dt>
703c703
< <em class="property">proc </em><code class="sig-name descname">!=</code><span class="sig-paren">(</span><em class="sig-param">const ref a: ?tset?</em>, <em class="sig-param">const ref b: tset?</em><span class="sig-paren">)</span>: bool<a class="headerlink" href="#Set.!=" title="Permalink to this definition">¶</a></dt>
---
> <em class="property">proc </em><code class="sig-name descname">!=</code><span class="sig-paren">(</span><em class="sig-param">const ref a: set(?t</em>, <em class="sig-param">?)</em>, <em class="sig-param">const ref b: set(t</em>, <em class="sig-param">?)</em><span class="sig-paren">)</span>: bool<a class="headerlink" href="#Set.!=" title="Permalink to this definition">¶</a></dt>
723c723
< <em class="property">proc </em><code class="sig-name descname">&lt;</code><span class="sig-paren">(</span><em class="sig-param">const ref a: ?tset?</em>, <em class="sig-param">const ref b: tset?</em><span class="sig-paren">)</span>: bool<a class="headerlink" href="#Set.<" title="Permalink to this definition">¶</a></dt>
---
> <em class="property">proc </em><code class="sig-name descname">&lt;</code><span class="sig-paren">(</span><em class="sig-param">const ref a: set(?t</em>, <em class="sig-param">?)</em>, <em class="sig-param">const ref b: set(t</em>, <em class="sig-param">?)</em><span class="sig-paren">)</span>: bool<a class="headerlink" href="#Set.<" title="Permalink to this definition">¶</a></dt>
743c743
< <em class="property">proc </em><code class="sig-name descname">&lt;=</code><span class="sig-paren">(</span><em class="sig-param">const ref a: ?tset?</em>, <em class="sig-param">const ref b: tset?</em><span class="sig-paren">)</span>: bool<a class="headerlink" href="#Set.<=" title="Permalink to this definition">¶</a></dt>
---
> <em class="property">proc </em><code class="sig-name descname">&lt;=</code><span class="sig-paren">(</span><em class="sig-param">const ref a: set(?t</em>, <em class="sig-param">?)</em>, <em class="sig-param">const ref b: set(t</em>, <em class="sig-param">?)</em><span class="sig-paren">)</span>: bool<a class="headerlink" href="#Set.<=" title="Permalink to this definition">¶</a></dt>
763c763
< <em class="property">proc </em><code class="sig-name descname">&gt;</code><span class="sig-paren">(</span><em class="sig-param">const ref a: ?tset?</em>, <em class="sig-param">const ref b: tset?</em><span class="sig-paren">)</span>: bool<a class="headerlink" href="#Set.>" title="Permalink to this definition">¶</a></dt>
---
> <em class="property">proc </em><code class="sig-name descname">&gt;</code><span class="sig-paren">(</span><em class="sig-param">const ref a: set(?t</em>, <em class="sig-param">?)</em>, <em class="sig-param">const ref b: set(t</em>, <em class="sig-param">?)</em><span class="sig-paren">)</span>: bool<a class="headerlink" href="#Set.>" title="Permalink to this definition">¶</a></dt>
783c783
< <em class="property">proc </em><code class="sig-name descname">&gt;=</code><span class="sig-paren">(</span><em class="sig-param">const ref a: ?tset?</em>, <em class="sig-param">const ref b: tset?</em><span class="sig-paren">)</span>: bool<a class="headerlink" href="#Set.>=" title="Permalink to this definition">¶</a></dt>
---
> <em class="property">proc </em><code class="sig-name descname">&gt;=</code><span class="sig-paren">(</span><em class="sig-param">const ref a: set(?t</em>, <em class="sig-param">?)</em>, <em class="sig-param">const ref b: set(t</em>, <em class="sig-param">?)</em><span class="sig-paren">)</span>: bool<a class="headerlink" href="#Set.>=" title="Permalink to this definition">¶</a></dt>
diff -r doc-master/html/_sources/modules/packages/HDF5.rst.txt doc-pr/html/_sources/modules/packages/HDF5.rst.txt
111c111
< .. function:: proc writeArraysToHDF5Files(dirName: string, dsetNames: [] string, filenames: [] string, type eltType, param rank: int, data: [] eltTypeArrayWrapperrank, mode: Hdf5OpenMode) throws
---
> .. function:: proc writeArraysToHDF5Files(dirName: string, dsetNames: [] string, filenames: [] string, type eltType, param rank: int, data: [] ArrayWrapper(eltType, rank), mode: Hdf5OpenMode) throws
diff -r doc-master/html/_sources/modules/packages/OrderedSet.rst.txt doc-pr/html/_sources/modules/packages/OrderedSet.rst.txt
231c231
<    .. method:: proc isDisjoint(const ref other: eltTypeorderedSet?): bool
---
>    .. method:: proc isDisjoint(const ref other: orderedSet(eltType, ?)): bool
242c242
<    .. method:: proc isIntersecting(const ref other: eltTypeorderedSet?): bool
---
>    .. method:: proc isIntersecting(const ref other: orderedSet(eltType, ?)): bool
367c367
< .. function:: proc &=(ref lhs: ?torderedSet?, const ref rhs: torderedSet?)
---
> .. function:: proc &=(ref lhs: orderedSet(?t, ?), const ref rhs: orderedSet(t, ?))
diff -r doc-master/html/_sources/modules/standard/Heap.rst.txt doc-pr/html/_sources/modules/standard/Heap.rst.txt
174c174
< .. function:: proc =(ref lhs: ?theap?, ref rhs: theap?)
---
> .. function:: proc =(ref lhs: heap(?t, ?), ref rhs: heap(t, ?))
diff -r doc-master/html/_sources/modules/standard/List.rst.txt doc-pr/html/_sources/modules/standard/List.rst.txt
253c253
<    .. method:: proc ref extend(other: eltTypelist?p)
---
>    .. method:: proc ref extend(other: list(eltType, ?p))
639c639
< .. function:: proc =(ref lhs: ?tlist?, rhs: tlist?)
---
> .. function:: proc =(ref lhs: list(?t, ?), rhs: list(t, ?))
654c654
< .. function:: proc ==(a: ?tlist?, b: tlist?): bool
---
> .. function:: proc ==(a: list(?t, ?), b: list(t, ?)): bool
666c666
< .. function:: proc !=(a: ?tlist?, b: tlist?): bool
---
> .. function:: proc !=(a: list(?t, ?), b: list(t, ?)): bool
diff -r doc-master/html/_sources/modules/standard/Set.rst.txt doc-pr/html/_sources/modules/standard/Set.rst.txt
94c94
<    .. method:: proc init=(const ref other: ?tset?)
---
>    .. method:: proc init=(const ref other: set(?t, ?))
124c124
<    .. method:: proc isDisjoint(const ref other: eltTypeset?): bool
---
>    .. method:: proc isDisjoint(const ref other: set(eltType, ?)): bool
139c139
<    .. method:: proc isIntersecting(const ref other: eltTypeset?): bool
---
>    .. method:: proc isIntersecting(const ref other: set(eltType, ?)): bool
227c227
< .. function:: proc =(ref lhs: ?tset?, const ref rhs: tset?)
---
> .. function:: proc =(ref lhs: set(?t, ?), const ref rhs: set(t, ?))
242c242
< .. function:: proc |(const ref a: ?tset?, const ref b: tset?): set(t)
---
> .. function:: proc |(const ref a: set(?t, ?), const ref b: set(t, ?)): set(t)
254c254
< .. function:: proc |=(ref lhs: ?tset?, const ref rhs: tset?)
---
> .. function:: proc |=(ref lhs: set(?t, ?), const ref rhs: set(t, ?))
263c263
< .. function:: proc +(const ref a: ?tset?, const ref b: tset?): tset?
---
> .. function:: proc +(const ref a: set(?t, ?), const ref b: set(t, ?)): set(t, ?)
276c276
< .. function:: proc +=(ref lhs: ?tset?, const ref rhs: tset?)
---
> .. function:: proc +=(ref lhs: set(?t, ?), const ref rhs: set(t, ?))
285c285
< .. function:: proc -(const ref a: ?tset?, const ref b: tset?): set(t)
---
> .. function:: proc -(const ref a: set(?t, ?), const ref b: set(t, ?)): set(t)
297c297
< .. function:: proc -=(ref lhs: ?tset?, const ref rhs: tset?)
---
> .. function:: proc -=(ref lhs: set(?t, ?), const ref rhs: set(t, ?))
311c311
< .. function:: proc &(const ref a: ?tset?, const ref b: tset?): set(t)
---
> .. function:: proc &(const ref a: set(?t, ?), const ref b: set(t, ?)): set(t)
323c323
< .. function:: proc &=(ref lhs: ?tset?, const ref rhs: tset?)
---
> .. function:: proc &=(ref lhs: set(?t, ?), const ref rhs: set(t, ?))
338c338
< .. function:: proc ^(const ref a: ?tset?, const ref b: tset?): set(t)
---
> .. function:: proc ^(const ref a: set(?t, ?), const ref b: set(t, ?)): set(t)
350c350
< .. function:: proc ^=(ref lhs: ?tset?, const ref rhs: tset?)
---
> .. function:: proc ^=(ref lhs: set(?t, ?), const ref rhs: set(t, ?))
365c365
< .. function:: proc ==(const ref a: ?tset?, const ref b: tset?): bool
---
> .. function:: proc ==(const ref a: set(?t, ?), const ref b: set(t, ?)): bool
378c378
< .. function:: proc !=(const ref a: ?tset?, const ref b: tset?): bool
---
> .. function:: proc !=(const ref a: set(?t, ?), const ref b: set(t, ?)): bool
390c390
< .. function:: proc <(const ref a: ?tset?, const ref b: tset?): bool
---
> .. function:: proc <(const ref a: set(?t, ?), const ref b: set(t, ?)): bool
402c402
< .. function:: proc <=(const ref a: ?tset?, const ref b: tset?): bool
---
> .. function:: proc <=(const ref a: set(?t, ?), const ref b: set(t, ?)): bool
414c414
< .. function:: proc >(const ref a: ?tset?, const ref b: tset?): bool
---
> .. function:: proc >(const ref a: set(?t, ?), const ref b: set(t, ?)): bool
426c426
< .. function:: proc >=(const ref a: ?tset?, const ref b: tset?): bool
---
> .. function:: proc >=(const ref a: set(?t, ?), const ref b: set(t, ?)): bool
diff -r doc-master/rst/modules/packages/HDF5.rst doc-pr/rst/modules/packages/HDF5.rst
111c111
< .. function:: proc writeArraysToHDF5Files(dirName: string, dsetNames: [] string, filenames: [] string, type eltType, param rank: int, data: [] eltTypeArrayWrapperrank, mode: Hdf5OpenMode) throws
---
> .. function:: proc writeArraysToHDF5Files(dirName: string, dsetNames: [] string, filenames: [] string, type eltType, param rank: int, data: [] ArrayWrapper(eltType, rank), mode: Hdf5OpenMode) throws
diff -r doc-master/rst/modules/packages/OrderedSet.rst doc-pr/rst/modules/packages/OrderedSet.rst
231c231
<    .. method:: proc isDisjoint(const ref other: eltTypeorderedSet?): bool
---
>    .. method:: proc isDisjoint(const ref other: orderedSet(eltType, ?)): bool
242c242
<    .. method:: proc isIntersecting(const ref other: eltTypeorderedSet?): bool
---
>    .. method:: proc isIntersecting(const ref other: orderedSet(eltType, ?)): bool
367c367
< .. function:: proc &=(ref lhs: ?torderedSet?, const ref rhs: torderedSet?)
---
> .. function:: proc &=(ref lhs: orderedSet(?t, ?), const ref rhs: orderedSet(t, ?))
diff -r doc-master/rst/modules/standard/Heap.rst doc-pr/rst/modules/standard/Heap.rst
174c174
< .. function:: proc =(ref lhs: ?theap?, ref rhs: theap?)
---
> .. function:: proc =(ref lhs: heap(?t, ?), ref rhs: heap(t, ?))
diff -r doc-master/rst/modules/standard/List.rst doc-pr/rst/modules/standard/List.rst
253c253
<    .. method:: proc ref extend(other: eltTypelist?p)
---
>    .. method:: proc ref extend(other: list(eltType, ?p))
639c639
< .. function:: proc =(ref lhs: ?tlist?, rhs: tlist?)
---
> .. function:: proc =(ref lhs: list(?t, ?), rhs: list(t, ?))
654c654
< .. function:: proc ==(a: ?tlist?, b: tlist?): bool
---
> .. function:: proc ==(a: list(?t, ?), b: list(t, ?)): bool
666c666
< .. function:: proc !=(a: ?tlist?, b: tlist?): bool
---
> .. function:: proc !=(a: list(?t, ?), b: list(t, ?)): bool
diff -r doc-master/rst/modules/standard/Set.rst doc-pr/rst/modules/standard/Set.rst
94c94
<    .. method:: proc init=(const ref other: ?tset?)
---
>    .. method:: proc init=(const ref other: set(?t, ?))
124c124
<    .. method:: proc isDisjoint(const ref other: eltTypeset?): bool
---
>    .. method:: proc isDisjoint(const ref other: set(eltType, ?)): bool
139c139
<    .. method:: proc isIntersecting(const ref other: eltTypeset?): bool
---
>    .. method:: proc isIntersecting(const ref other: set(eltType, ?)): bool
227c227
< .. function:: proc =(ref lhs: ?tset?, const ref rhs: tset?)
---
> .. function:: proc =(ref lhs: set(?t, ?), const ref rhs: set(t, ?))
242c242
< .. function:: proc |(const ref a: ?tset?, const ref b: tset?): set(t)
---
> .. function:: proc |(const ref a: set(?t, ?), const ref b: set(t, ?)): set(t)
254c254
< .. function:: proc |=(ref lhs: ?tset?, const ref rhs: tset?)
---
> .. function:: proc |=(ref lhs: set(?t, ?), const ref rhs: set(t, ?))
263c263
< .. function:: proc +(const ref a: ?tset?, const ref b: tset?): tset?
---
> .. function:: proc +(const ref a: set(?t, ?), const ref b: set(t, ?)): set(t, ?)
276c276
< .. function:: proc +=(ref lhs: ?tset?, const ref rhs: tset?)
---
> .. function:: proc +=(ref lhs: set(?t, ?), const ref rhs: set(t, ?))
285c285
< .. function:: proc -(const ref a: ?tset?, const ref b: tset?): set(t)
---
> .. function:: proc -(const ref a: set(?t, ?), const ref b: set(t, ?)): set(t)
297c297
< .. function:: proc -=(ref lhs: ?tset?, const ref rhs: tset?)
---
> .. function:: proc -=(ref lhs: set(?t, ?), const ref rhs: set(t, ?))
311c311
< .. function:: proc &(const ref a: ?tset?, const ref b: tset?): set(t)
---
> .. function:: proc &(const ref a: set(?t, ?), const ref b: set(t, ?)): set(t)
323c323
< .. function:: proc &=(ref lhs: ?tset?, const ref rhs: tset?)
---
> .. function:: proc &=(ref lhs: set(?t, ?), const ref rhs: set(t, ?))
338c338
< .. function:: proc ^(const ref a: ?tset?, const ref b: tset?): set(t)
---
> .. function:: proc ^(const ref a: set(?t, ?), const ref b: set(t, ?)): set(t)
350c350
< .. function:: proc ^=(ref lhs: ?tset?, const ref rhs: tset?)
---
> .. function:: proc ^=(ref lhs: set(?t, ?), const ref rhs: set(t, ?))
365c365
< .. function:: proc ==(const ref a: ?tset?, const ref b: tset?): bool
---
> .. function:: proc ==(const ref a: set(?t, ?), const ref b: set(t, ?)): bool
378c378
< .. function:: proc !=(const ref a: ?tset?, const ref b: tset?): bool
---
> .. function:: proc !=(const ref a: set(?t, ?), const ref b: set(t, ?)): bool
390c390
< .. function:: proc <(const ref a: ?tset?, const ref b: tset?): bool
---
> .. function:: proc <(const ref a: set(?t, ?), const ref b: set(t, ?)): bool
402c402
< .. function:: proc <=(const ref a: ?tset?, const ref b: tset?): bool
---
> .. function:: proc <=(const ref a: set(?t, ?), const ref b: set(t, ?)): bool
414c414
< .. function:: proc >(const ref a: ?tset?, const ref b: tset?): bool
---
> .. function:: proc >(const ref a: set(?t, ?), const ref b: set(t, ?)): bool
426c426
< .. function:: proc >=(const ref a: ?tset?, const ref b: tset?): bool
---
> .. function:: proc >=(const ref a: set(?t, ?), const ref b: set(t, ?)): bool
</pre>
</details>

resolves #13626

cassella commented:
(force pushed after setting up new laptop with DCO hook.)

cassella commented:
(force push after remembering that identifiers can start with `$`.

cassella commented:
sigh, force push after actually testing my memory and finding `$` can't be the first character of an identifier after all.)

lydia-duncan commented:
Dear god, that's an embarrassing number of cases that were improved.  I posted with an idea on the original issue, hopefully that works in a way that avoids these additional edge cases - we ought to be able to walk through the list of primitives to see viable candidates instead of having to go through the changed output to find potential issues (though we definitely should walk through the changed output).

cassella commented:
I don't actually know how to AST, so I don't expect to be able to come up with a more principled approach in the near term.  I was suspicious of these lines (since they're concatenating three things, which is what the bad behavior looked like here), but I don't understand any of the logic involved in reaching them.  I'd probably need a lot of time and guidance to make any conceptual improvement to this PR.  I won't take offense if someone who knows what they're doing starts over with a better approach.

It occurs to me that if this function is used outside of chpldoc (e.g., in producing error messages), I should probably have tested more broadly as well.


Where I'm at in the meantime, is this code is in the block that opens `appendExpr()`: `if (expr->primative == 0)`.  Should I still be able to find some part of the expr to check for the `PRIM_` values?  I'll try to figure that out.

lydia-duncan commented:
> I don't actually know how to AST, so I don't expect to be able to come up with a more principled approach in the near term.

That's fair.  The trouble is I'm not sure this is going to be viewed as a priority for me, so I don't know if I will be able to do more than review right now. :/

cassella commented:
Writing `proc foo(bar = x + y)` produces `foo(bar = x+y)` without tripping code I added there to trip on `expr->isPrimitive(PRIM_ADD)`.  That plus the tuple-handling code using a string match on `*` with the comment:

```c
          // I found that some multiplies would match this (even though they
          // really should be PRIM_MULT),
```

is making me think the PRIM_ values aren't available here at all.


With the string compare-based approach, I think I could make the code a little cleaner by pulling out the two-args and `!isalpha() && != "_"` part into a new `looksLikeOperator() function, so the flow could be

```c++
    ... else if ((char *name = looksLikeOperator()) != NULL) {
        if (!strcmp(name, "*")) {
            use tuple format
        } else {
            use infix
        }
    } else {
        the existing appendExpr(expr, fname, printingType);
    }
```

instead of having separate 2-arg and not-2-arg cases that do the same thing.


But any more substantial improvement to the PR is probably beyond me.  Even the suggestion to walk through the list of primitives I don't understand.  :(

lydia-duncan commented:
> the PRIM_ values aren't available here at all

Ah, that's a shame, but I trust your check.

> I think I could make the code a little cleaner by pulling out the two-args and `!isalpha() && != "_"` part into a new `looksLikeOperator() function

That seems reasonable to me.

> Even the suggestion to walk through the list of primitives I don't understand

Since the primitives don't seem to be relevant to chpldoc during that pass, it's a moot point anyways :)

cassella commented:
> Ah, that's a shame, but I trust your check.

Well, I wouldn't.  :)  But since this passes all the chpldoc tests, and adds a test to lock the improvements in (below), I think it can at least be a placeholder for any future conceptual improvement.


While I was doing the cleanup, I noticed the code at the top of the function for handling unary operators was missing two, so I fixed those also.

While I was there, I noticed the `if()` at the beginning of the function, containing (eventually) the binary-operator code, already had local variables `sel` and `fnName` for the unresolved sym expr and its unresolved name, so I used them instead of re-obtaining them,

```chapel
    if (UnresolvedSymExpr* sel = toUnresolvedSymExpr(expr->baseExpr))
    {
      const char* fnName = sel->unresolved;
```

I added a test that uses all the operators in default arg expressions, as  well as a few different number-of-arguments functions.  The test passes with this PR; the failing cases on master are:

```
 .. function:: proc bitwiseNegDefault(arg = ~(x))
 .. function:: proc notNilDefault(arg = postfix!(c))
 .. function:: proc twoArgFuncDefault(arg = xf2y)
```

I added futures for #16939, #16940, #16941 as well.  I put them in this PR since they depend on the `CHPLDOCOPTS` and `CLEANFILES` files that the test that belongs in this PR puts in the directory.

Also I squashed everything but those three futures into one commit, since the initial commit wasn't very involved, is completely different after the rework, and isn't worth preserving in the history.

In retrospect, and after seeing the same behavior with `proc foo(arg=try! x)` as `var q = try! x` in #16866, probably none of these have anything to do with being a function arg's default value expression, but just being the representation of that expression in any form.  I could rewrite the tests as the latter form, if that would be better.  (And put them somewhere else.)

lydia-duncan commented:
Thanks for doing all that!  I'll give this a proper review soon so we can merge it.

>  I could rewrite the tests as the latter form, if that would be better. (And put them somewhere else.)

I wouldn't get rid of the tests in their current form, but I wouldn't object to having a copy in the other form.  chpldoc tests are really fast to run because they skip so much of other compilation, and I personally believe having coverage is a good thing, even if the implementation is such that they won't have different results - a future code change could cause them to diverge and we wouldn't notice without such tests.

lydia-duncan commented:
While I'm reviewing, does the PR description need updating for your changes?

lydia-duncan commented on compiler/AST/AstToText.cpp:
```diff
@@ -1135,12 +1163,11 @@ void AstToText::appendExpr(CallExpr* expr, bool printingType)
       {
-        UnresolvedSymExpr* name     = toUnresolvedSymExpr(expr->baseExpr);
-        if (printingType && strcmp(name->unresolved, "*") == 0)
+        // Lydia 2015/02/17 ... except homogeneous tuple inner workings.
```
I think it's fair to remove my name and that date from the comment and just let it talk about homogeneous tuples on their own

cassella replied:
I'll update the PR message and that comment, and add a test of the `var a = something;` forms.  I'll try to get to it tonight.

cassella replied:
I did those things, and also added a third variant of the test where the things being set to different expressions are record members.

Squashed again.

I also noticed that a few of the function arg tests I'd written as `proc foo(args);` instead of `proc foo(args) {}`; I fixed those.

I also noticed that in the variants where `something = x.type`, chpldoc emits an extra space after the `x.type` in its output, visible easily in the function arg .good file since there's a `)` following, and less easily in an editor in the other two.  E.g.,

```
.. function:: proc tuple3Default(arg = (x, y, z))
.. function:: proc dotTypeDefault(arg = x.type )
```

I can file a new issue and future for that, but probably not tonight.

lydia-duncan reviewed (commented):
Just a minor adjustment to a comment required, and that question I had about the PR message.  I'll run testing over it now (since neither of those should impact behavior)

cassella commented:
Hmm, I'd also included `init=` in `looksLikeInfixOperator()`, but now I think that was wrong.  `init=` replaces the `=`; it's not going to turn up on the rhs of an `=` (once/if it's implemented).

And I'm also realizing none of my tests cover the `printingType=true` case, such as the original `x: map(k, v)` sort of thing.  I'll add some of those too.

cassella commented:
Done.  This time I didn't squash, for ease of incremental review.  I can squash before it's merged.

I added a case to each test of `var x = c!.aField`.  Unlike `var x = x.type`, there's no trailing space added.

I added a set of cases to each test of `var x: R(*)` and `var X:R(*) = new R(*)` for generic records `R` having 0, 1, 2, and 3 generic (param) fields.  All look good.

Incidentally, something's a little surprising about the documentation of those generic records, e.g.,:

```chapel
  record R3 { param a,b,c: int; }
```

```
+.. record:: R3
+
+   .. attribute:: param a: int
+
+   .. attribute:: param b: a.type 
+
+   .. attribute:: param c: b.type 
```

I'm a bit surprised `b`'s type is given as `a.type` instead of `int`, though I recall that's how the spec describes it.  So I don't think this is a bug, just surprising.

(And the lines for `b` and `c` come out with a trailing space.)

`make docs` produces the same `doc/` directory as the one compared to master's in the  first ccomment.

lydia-duncan commented:
> This time I didn't squash, for ease of incremental review. I can squash before it's merged.

Reviewed, thanks!  I don't think there's a need to squash again before merge - I'd just feel the need to run testing over it again in that case, and I don't think this branch is too complicated or bloated or anything.  I've added "resolves #13626" to allow that issue to be closed when this is merged - let me know if you think that's the wrong call

I'll probably wait to merge until Tuesday cuz of the holiday and weekend, and am running one last sanity check on my end, but I think this is good to go.  Thanks for handling it :)

cassella commented:
>  I've added "resolves #13626" 

Ah, thanks!  I'd put that in the first comment for some reason.

Sounds good!
